git-pr-merge

```diff
@@ -65,40 +87,114 @@ prmerge::setup() {
   fi
 }
 
-prmerge::prepare() {
-  echo merging "${feature}" to "${master}"
-
-  # Fetch from all remotes so we can check if local branch is up-to-date
-  git remote update >/dev/null
-
-  if [[ "$(git rev-parse "${master}")" != "$(git rev-parse "${master}@{upstream}")" ]]; then
-    printf '%s is not up-to-date. You should run:\n' "${master}" >&2
-    printf '  git checkout %s\n' "${master}" >&2
-    printf '  git pull\n' >&2
-    return 1
-  fi
-
+common::prepare() {
   # cd to root of repo as master may not contain CWD, but switch back
   # when we're done.
   # shellcheck disable=SC2064
   # We want to evaluate $(pwd) now and not later.
   trap "cd \"$(pwd)\"" EXIT
   cd "$(git rev-parse --show-toplevel)"
 
+  # Fetch from all remotes so we can check if local branch is up-to-date
+  git remote update >/dev/null
+}
+
+prmerge::prepare() {
+  common::prepare
+
+  local do_pull=false
+  if [[ "$(git rev-parse "${master}")" != "$(git rev-parse "${master}@{upstream}")" ]]; then
+    printf '\n%s is not up-to-date. To update, press enter (^C to abort):' "${master}"
+    read -r
+    do_pull=true
+  fi
+
   git checkout -q "${master}"
+
+  if "${do_pull}"; then
+    git pull
+    if ! git merge-base --is-ancestor "${master}" "${feature}"; then
+      git checkout "${feature}"
+      printf 'Local %s is now up-to-date. Update %s before merging\n' "${master}" "${feature}"
+      exit 1
+    fi
+  fi
+}
+
+prupdate::prepare() {
+  common::prepare
+
+  if git merge-base --is-ancestor "${master}@{upstream}" "${feature}"; then
+    printf 'Already up to date\n'
+    exit 0
+  fi
+
+  # We need to ensure the PR base.sha is up-to-date, otherwise we create a
+  # merge commit on the PR that dismisses approval. This can be done by
+  # force pushing a commit on the PR feature branch, but we need to make sure
```

Comment:
doesn't need to be a force push I believe - ie create merge commit, push it (no force required) that will update base.ref. the undoing requires force.

also you seem to have removed comments re undoing previous merges. if block.
maybe the comment above should go above the else block and the if block should have its own?

Reply:
I think I determined it was the force push that actually updated the base.sha hash. Pushing the foxtrot merge was not enough.

I did remove that comment, because the printf text says it. It seemed redundant as a comment. The block comment applies to both. in both cases we need a force push to update the base.sha

Reply:
ok.

Reply:
I conceded to this before, but I'm actually not sure about this now.
I don't think it was the forcepush that made it happen - is a forcepush that could be an ordinary push even any different?
anyway - my rationale was: let's try and do what github does when clicking "update branch" which is a merging master into feature. I do not think a force-push is required for this. if you want we can test it out again.

Reply:
I'm not 100% sure either. I'll retest. I remember expecting the base hash to update when I pushed them merge commit but vaguely remember being surprised that it was not, but we tried so many things I could easily have it wrong.

Reply:
Also, I don't do force push to do the github foxtrot merge - that's a normal push. The force push is undoing it

Reply:
We seem to be talking past one another a little.

1) "let's try and do what github does when clicking "update branch" which is a merging master into feature. I do not think a force-push is required for this" - that's correct. That's what I'm doing. The force push is the subsequent push to remove the merge of master into feature.

2) I have re-tested it and you are right. Pushing the foxtrot merge is enough to get the base.sha of the PR updated. We previously demonstrated that doing this does not remove approval.

I will update the comments to be accurate WRT this.